pydantic/main.py

```diff
@@ -362,18 +365,42 @@ def copy(
 
     @classmethod
     def model_json_schema(
-        cls, by_alias: bool = True, ref_template: str = default_ref_template
+        cls,
+        by_alias: bool = True,
+        ref_template: str = DEFAULT_REF_TEMPLATE,
+        schema_generator: type[GenerateJsonSchema] = GenerateJsonSchema,
     ) -> typing.Dict[str, Any]:
+        """
+        To override the logic used to generate the JSON schema, you can create a subclass of GenerateJsonSchema
+        with your desired modifications, then override this method on a custom base class and set the default
+        value of `schema_generator` to be your subclass.
+        """
         cached = cls.__schema_cache__.get((by_alias, ref_template))
         if cached is not None:
             return cached
-        s = model_schema(cls, by_alias=by_alias, ref_template=ref_template)
+        s = schema_generator(by_alias=by_alias, ref_template=ref_template).generate(cls.__pydantic_core_schema__)
         cls.__schema_cache__[(by_alias, ref_template)] = s
         return s
 
+    @classmethod
+    def model_json_schema_misc(cls) -> JsonSchemaMisc | None:
```

Comment:
I don't love making this a separate method from `model_json_schema`, BUT it kind of serves a different purpose, and reduces the amount of boilerplate that would be necessary to make many simple changes to the schema.

Open to renaming and/or replacing this method with a different way of achieving the same result if preferred.

Reply:
Perhaps I should be trying to do more with `__modify_schema__`...